dashboard/test/models/pd/application/application_base_test.rb

```diff
@@ -18,24 +18,34 @@ class ApplicationBaseTest < ActiveSupport::TestCase
           'Application type is not included in the list',
           'Application year is not included in the list',
           'Type is required'
-        ],
-        application.errors.full_messages
+        ].sort,
+        application.errors.full_messages.sort
       )
     end
 
     test 'derived classes override type and year' do
-      application = TEACHER_APPLICATION_CLASS.new
+      application = create TEACHER_APPLICATION_FACTORY
       assert_equal TEACHER_APPLICATION, application.application_type
       assert_equal APPLICATION_CURRENT_YEAR, application.application_year
     end
 
     test 'default status is unreviewed' do
       application = ApplicationBase.new
+      application.save
```

Comment:
I originally used `application = create TEACHER_APPLICATION_FACTORY` instead, without a save. Wanted to stick to using ApplicationBase. I have no idea why this way works––status is being set `before_validation`.

I don't know why other validations aren't failing ("Form data is required, Type is required, User is required, Application type is not included in the list, Application year is not included in the list"). According to https://guides.rubyonrails.org/getting_started.html#validations-and-displaying-error-messages, validations happen before the object is saved. Why would the `before_validation` be triggered but not the other validations?

Reply:
Ah, thank you. Well, it's nice to avoid the validations causing a failure in this case tehe. Or we could create using `TEACHER_APPLICATION_FACTORY` and not need a save (was trying to avoid that to stay inside of ApplicationBase) ... do you have a strong preference?

Reply:
I am not sure I understand the situation. why did you need to add a save (or anything else) here?

if the situation is that this test is suddenly failing, i would say that, it doesn't really make sense to me to be instantiating the ApplicationBase class here, because that's not something we ever need to do in our application -- we are only every instantiating children of this class (please correct me if I'm wrong there). based on that, here are my thoughts:
1.  instantiate teacher application here instead
2.  should other classes which extend ApplicationBase should also have similar tests?
3.  should these tests move into the inheriting class's test? e.g. teacher_application_test.rb.

based on the pattern existing pattern here, I'd suggest doing (1), not doing (3), and poking around a bit to see if (2) seems necessary based on the changes you are making.

Reply:
Changed it back to (1). As for (2):
- `teacher_application_test.rb` has a test making sure we can go from unreviewed to pending
- there are tests that shouldn't work in both `facilitator1819_application_test.rb` and `facilitator1819_application_test.rb` but they are working for different reasons (mostly involving the "locked" state of the application being tested correctly and overriding whatever's happening with the "status")
- `principal_approval_application_test.rb` has a test making sure status can be nil, but the test feels a bit wonky because the status is never actually nil (e.g. in staging, the status is set as unreviewed upon `build` even though we say `status: nil` in the test)

So, options from here:
1) I refactor tests from all these files before shipping out this PR, and add tests explicitly making sure the functionality is consistent from before
2) I add a separate ticket to refactor these (will it ever get picked up?), as the tests don't warrant a refactor based on the changes I'm making, but rather on how we're currently testing 
I *think* (1) would take about half a day––I'd mostly manually update the status after creating the application.

Reply:
It is unclear for me what the default status should be for the principal application. The tests indicate "nil" status by default (and we have a test in application_base for that case), but it looks like application_base writes it as "unreviewed" (in staging). FacilitatorApplication seems clear that it is "unreviewed" by default, although there isn't a strong test for it right now.

I don't even think we do facilitator applications any more. It's unclear to me how much time to invest in making sure those work––not sure best practice here

Reply:
ok, thanks for looking, I think it is fine to just change the unit test to create a teacher application.